### PR TITLE
wallet: Fix issues when `walletdir` is root directory

### LINKED ISSUE #21510: Special characters in `wallet_name`
I am testing few things with different strings in `wallet_name` containing special characters because of few related issues and PRs mentioned in the end.

**Expected behavior**

Ideally few special chars should not be allowed and if allowed the wallet should be created only in wallets directory unless a path is mentioned with other argument. Right now same argument is used for name and path. Maybe there can be better solutions but Bitcoin Core's weird behavior with wallet names in some cases can be fixed.

**Actual behavior**

I tried creating few wallets with random strings as `wallet_name`. Some of them are not even visible in `listwalletdir` or _Bitcoin-Qt->File->Open Wallet_

**To reproduce**

```python
import random 
import string
from bitcoinrpc.authproxy import AuthServiceProxy

rpc_port = 18444
rpc_user = 'user3'
rpc_password = 'pass3'

def wallet_name(size): 
	
	generate_wallet = ''.join([random.choice(string.punctuation + string.ascii_letters) 
						 for n in range(size)]) 
						
	return generate_wallet
 

rpc_connection = AuthServiceProxy("http://%s:%s@127.0.0.1:%s"%(rpc_user,rpc_password,rpc_port))

for i in range(100):
    wallet = wallet_name(20)
    print(wallet)
    try:
        result = rpc_connection.createwallet(wallet)        
    except:
        pass    
```

The above code will try to create 100 wallets with random strings containing special characters. Not every string tried (and printed) will be used for creating a wallet, exceptions are ignored and next string is tried. 

Once few wallets are created with random strings, we can check the wallet folders in `testnet3\wallets` which is different from the wallets that we see in results of `listwalletdir` or _Bitcoin-Qt->File->Open Wallet_

**System information**

Bitcoin Core v0.21.0

Windows 10

Related issues: https://github.com/bitcoin/bitcoin/issues/21501 https://github.com/bitcoin/bitcoin/issues/19928 https://github.com/bitcoin/bitcoin/issues/20128
Related PRs: https://github.com/bitcoin/bitcoin/pull/20080 https://github.com/bitcoin/bitcoin/pull/20393 https://github.com/bitcoin/bitcoin/pull/20741

### PR DESCRIPTION
**Pull request by ghost**
+ Remove one character less from wallet path

+ After testing lot of random strings with special chars in `wallet_name`, I found that the issue was not related to special characters in the name. Reviewing PR https://github.com/bitcoin/bitcoin/pull/21907 helped me resolve the issue.

**Real issue**: If the path mentioned in `walletdir` is a root directory, first character of the wallet name or path is removed

**Solution**: `if` statement to check `walletdir` is a root directory

Fixes: https://github.com/bitcoin/bitcoin/issues/21510 https://github.com/bitcoin/bitcoin/issues/21501 
Related PR: https://github.com/bitcoin/bitcoin/pull/20080

Consider the wallet directories `w1` and `w2` saved in `D:\`. Run `bitcoind.exe -walletdir=D:\`, Results for `bitcoin-cli.exe listwalletdir`:

Before this PR:

```

{
  "wallets": [
    {
      "name": "1"
    },
    {
      "name": "2"
    }
  ]
}
 
```

After this PR:
```
  "wallets": [
    {
      "name": "w1"
    },
    {
      "name": "w2"
    }
  ]
}
 
```

### CODE REVIEW

**Comment by Talkless** [comment]
Welp, I did crazy thing and launched `bitcoind` as root, with:
```
sudo src/bitcoind -regtest -walletdir=/ -datadir=/tmp/datadir -server
```
Created wallet, and then listed:
```
{
  "wallets": [
    {
      "name": "est"
    },
    {
      "name": "ome/vincas/.bitcoin/regtest/wallets"
    },
    {
      "name": "ome/vincas/code/eclair/eclair_recovery_tool.git/eclair-core/target/integration-8198c6f7-2f0b-4394-a92a-d4d738994f33/datadir-bitcoin/regtest/wallets"
    },
    {
      "name": "ome/vincas/code/eclair/eclair_recovery_tool.git/eclair-core/target/integration-9063e6a5-c013-4127-a7dd-a10a3874ed18/datadir-bitcoin/regtest/wallets"
    },
    {
      "name": "ome/vincas/code/eclair/eclair_recovery_tool.git/eclair-core/target/integration-e9783f5b-51ab-458f-ab57-31976e38bff8/datadir-bitcoin/regtest/wallets"
    },
    {
      "name": "ome/vincas/code/eclair/eclair_recovery_tool.git/eclair-core/target/integration-f1c34eba-d63e-4379-a315-75056bbcc162/datadir-bitcoin/regtest/wallets"
    },
    {
      "name": "ome/vincas/code/eclair/eclair_recovery_tool.git/eclair-core/target/integration-d4096df3-5adb-40dd-9425-6f80fb6fd07b/datadir-bitcoin/regtest/wallets"
    },
    {
      "name": "ome/vincas/code/eclair/eclair_recovery_tool.git/eclair-core/target/integration-e6a27da9-a78e-48a5-a272-38a950ddfd41/datadir-bitcoin/regtest/wallets"
    },
    {
      "name": "ome/vincas/code/bitcoin/123-bitcoin-core-gui.git/test/functional/data/wallets/high_minversion"
    },
    {
      "name": "ome/vincas/code/bitcoin/149-bitcoin-core-gui.git/test/functional/data/wallets/high_minversion"
    }
  ]
}
```

Interesting to see `bitcoind` scan whole system for wallets :D (log even printed `ListDatabases: Error scanning /sys/fs/bpf: boost::filesystem::status: Operation not permitted: "/sys/fs/bpf/wallet.dat"`).

Anyhow, issue is the same on Linux, and this PR did fix that too:
```
{
  "wallets": [
    {
      "name": "test"
    },
...same...
```

File: src/wallet/db.cpp [hunk 1]
```diff
@@ -25,7 +25,13 @@ std::vector<fs::path> ListDatabases(const fs::path& wallet_dir)
         try {
             // Get wallet path relative to walletdir by removing walletdir from the wallet path.
             // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
-            const fs::path path = it->path().string().substr(offset);
+            fs::path path = wallet_dir.string();
```
Comment (Talkless) [inline src/wallet/db.cpp:28]: You could still keep `path` const as before, by utilizing immediately-invoked lambda, as suggested in [ES.28: Use lambdas for complex initialization, especially of const variables](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init) guideline, with something like this (also avoiding re-initializing `path` variable):

```
            const fs::path path = [&wallet_dir, &it, offset]() {

                if (wallet_dir == wallet_dir.root_path()) {
                   return it->path().string().substr(offset - 1);
                }

                return it->path().string().substr(offset);
            }();
```
Not sure if this is the best way to do that overall, as I don't have basically any experience with boost/stl filesystem...

**Comment by prayank23** [comment]
> Interesting to see bitcoind scan whole system for wallets :D (log even printed ListDatabases: Error scanning /sys/fs/bpf: boost::filesystem::status: Operation not permitted: "/sys/fs/bpf/wallet.dat").

Yes it recursively iterates path mentioned. How much time did this take?

> Anyhow, issue is the same on Linux, and this PR did fix that too

Interesting. Thanks for testing this on Linux. I will change name, description, commit message and comment after testing few more things.

Although I think it's less likely someone will save wallet in root on Linux. But it's normal to save different folders in root of different drives on Windows.

File: src/wallet/db.cpp (hunk 1 above)
Comment (prayank23) [reply src/wallet/db.cpp:28]: Will try this approach

File: src/wallet/db.cpp [hunk 2]
```diff
@@ -25,7 +25,13 @@ std::vector<fs::path> ListDatabases(const fs::path& wallet_dir)
...
+
+            if (path == path.root_path()) { //If path is a 'root directory' for Windows
+               path = it->path().string().substr(offset - 1);
```
Comment (promag) [inline src/wallet/db.cpp:31]: nit, wrong indent

**Review by promag — COMMENTED** [review]
Not sure it's worth adding these cases considering #20744 unless we want to backport the fix.

**Comment by Talkless** [comment]
> Yes it recursively iterates path mentioned. How much time did this take?

About 19s in a VM running on i7 laptop with SATA SSD.

**Comment by ryanofsky** [comment]
Definitely good to have a fix for this issue. Would welcome simplifying the PR though. I think the following would be a simpler short term fix than c5ee096d596766b5f9dd6e3f8d81c8e87c0e9c3a:

```diff
diff --git a/src/wallet/db.cpp b/src/wallet/db.cpp
index cd49baeb786..993dd09b8b9 100644
--- a/src/wallet/db.cpp
+++ b/src/wallet/db.cpp
@@ -12,7 +12,7 @@
 
 std::vector<fs::path> ListDatabases(const fs::path& wallet_dir)
 {
-    const size_t offset = wallet_dir.string().size() + 1;
+    const size_t offset = wallet_dir.string().size() + (wallet_dir == wallet_dir.root_name() ? 0 : 1);
     std::vector<fs::path> paths;
     boost::system::error_code ec;
 
```

This:

- avoids adding and then subtracting 1 unnecessarily
- avoids doing the same comparison in a loop
- avoids using an unnecessary lambda like
- uses root_name instead of root_path which should avoid triggering for `/` on linux according to https://www.boost.org/doc/libs/1_69_0/libs/filesystem/doc/reference.html#path-decomposition-table

In long term after we switch from boost (#20744) or update boost (#19667), a more robust fix also suggested https://github.com/bitcoin/bitcoin/issues/21501#issuecomment-807122155 would be:

```diff
--- a/src/wallet/db.cpp
+++ b/src/wallet/db.cpp
@@ -12,7 +12,6 @@
 
 std::vector<fs::path> ListDatabases(const fs::path& wallet_dir)
 {
-    const size_t offset = wallet_dir.string().size() + 1;
     std::vector<fs::path> paths;
     boost::system::error_code ec;
 
@@ -24,8 +23,7 @@ std::vector<fs::path> ListDatabases(const fs::path& wallet_dir)
 
         try {
             // Get wallet path relative to walletdir by removing walletdir from the wallet path.
-            // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
-            const fs::path path = it->path().string().substr(offset);
+            const fs::path path = fs::lexically_relative(it->path(), wallet_dir);
 
             if (it->status().type() == fs::directory_file &&
                 (IsBDBFile(BDBDataFile(it->path())) || IsSQLiteFile(SQLiteDataFile(it->path())))) {
```

If we wanted to be really ambitious maybe we could write a unit test for the bug using [subst](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/subst) `subprocess.run(f"subst {tempdir} x:")`, `node.start(["-walletdir=x:"])`

**Comment by prayank23** [comment]
Thanks for the review @ryanofsky 

Made changes in https://github.com/bitcoin/bitcoin/commit/d44a261acff40c1c8727d3cc0106bde65a6416d0

> Would welcome simplifying the PR though. I think the following would be a simpler short term fix than c5ee096

Agree

**Review by ryanofsky — APPROVED** [review]
Code review ACK d44a261acff40c1c8727d3cc0106bde65a6416d0

For background on this issue: the problem of wallet name truncation would have started happening for `X:` style wallet dirs after #14561 and would have started happening for `X:\` style wallet dirs after #20080, IIUC.

Long term fix should be switching `lexically_relative` as described https://github.com/bitcoin/bitcoin/pull/21944#issuecomment-850023110

**Review by meshcollider — COMMENTED** [review]
utACK d44a261acff40c1c8727d3cc0106bde65a6416d0

**Pull request merged**